Consider making `defaults.plugins` TypeScript type non-Partial

When I updated from Chart.js 3.0.0 beta 7 to RC 3, I started getting type errors similar to the following in my code:

```
src/legendUtils.ts:39:20 - error TS2532: Object is possibly 'undefined'.

39   const original = defaults.plugins.legend.labels.generateLabels;
                      ~~~~~~~~~~~~~~~~~~~~~~~
```

I saw that the type definition for `plugins` was changed in 49b0916 to use `Partial`.

If I'm understanding correctly:

1. If a plugin isn't registered, then its corresponding `defaults.plugins` property is undefined / not set.
2. If a plugin is registered, then its corresponding `defaults.plugins` property is always set.
3. If a plugin is in use, then setting its corresponding `defaults.plugins` property to undefined stands a good chance of breaking things.

Letting `defaults.plugins` be Partial accurately reflects item 1, but it means that any code that references it has to do non-nullable assertions (which goes against common ESLint configurations) or optional chaining operators every time it accesses it (even though item 2 says that it's guaranteed to be set, as long as `register` is correctly called), and it allows you to set the defaults.plugins to undefined (which would break item 3).

Weighing all of that, I think it makes the most sense to mark `defaults.plugins` properties as required (in spite of item 1). To do this, I believe you could change the type of `PluginChartOptions` itself to not use `Partial` (since most places that reference it are already using `DeepPartial`), or you could change `Defaults` to extend `Required<PluginChartOptions<ChartType>>`.

Feedback / disagreement / etc. welcome. Thank you.

--- a/tsconfig.json
+++ b/tsconfig.json
@@ -20,5 +20,8 @@
   "include": [
     "./src/**/*.js",
     "./types"
+  ],
+  "exclude": [
+    "./types/tests"
   ]
 }

--- a/types/index.esm.d.ts
+++ b/types/index.esm.d.ts
@@ -644,14 +644,13 @@ export interface Defaults extends CoreChartOptions<ChartType>, ElementChartOptio
 }
 
 export type Overrides = {
-  [key in ChartType]: DeepPartial<
+  [key in ChartType]:
     CoreChartOptions<key> &
     ElementChartOptions &
     PluginChartOptions<key> &
     DatasetChartOptions<ChartType> &
     ScaleChartOptions<key> &
-    ChartTypeRegistry[key]['chartOptions']
-    >;
+    ChartTypeRegistry[key]['chartOptions'];
 }
 
 export const defaults: Defaults;
@@ -2572,7 +2571,7 @@ export interface PluginOptionsByType<TType extends ChartType> {
   tooltip: TooltipOptions<TType>;
 }
 export interface PluginChartOptions<TType extends ChartType> {
-  plugins: Partial<PluginOptionsByType<TType>>;
+  plugins: PluginOptionsByType<TType>;
 }
 
 export interface GridLineOptions {
@@ -3257,9 +3256,9 @@ export interface ChartTypeRegistry {
 
 export type ChartType = keyof ChartTypeRegistry;
 
-export type ScaleOptionsByType<TScale extends ScaleType = ScaleType> = DeepPartial<
+export type ScaleOptionsByType<TScale extends ScaleType = ScaleType> =
   { [key in ScaleType]: { type: key } & ScaleTypeRegistry[key]['options'] }[TScale]
->;
+;
 
 export type DatasetChartOptions<TType extends ChartType = ChartType> = {
   [key in TType]: {

--- a/types/tests/plugins/defaults.ts
+++ b/types/tests/plugins/defaults.ts
@@ -1,0 +1,18 @@
+import { defaults } from '../../index.esm';
+
+// https://github.com/chartjs/Chart.js/issues/8711
+const original = defaults.plugins.legend.labels.generateLabels;
+
+defaults.plugins.legend.labels.generateLabels = function(chart) {
+  return [{
+    datasetIndex: 0,
+    text: 'test'
+  }];
+};
+
+// @ts-expect-error Type '{ text: string; }[]' is not assignable to type 'LegendItem[]'.
+defaults.plugins.legend.labels.generateLabels = function(chart) {
+  return [{
+    text: 'test'
+  }];
+};

--- a/types/tests/scales/options.ts
+++ b/types/tests/scales/options.ts
@@ -1,0 +1,32 @@
+import { Chart } from '../../index.esm';
+
+const chart = new Chart('test', {
+  type: 'bar',
+  data: {
+    labels: ['a'],
+    datasets: [{
+      data: [1],
+    }, {
+      type: 'line',
+      data: [{ x: 1, y: 1 }]
+    }]
+  },
+  options: {
+    scales: {
+      x: {
+        type: 'time',
+        time: {
+          unit: 'year'
+        }
+      },
+      x1: {
+        // @ts-expect-error Type '"linear"' is not assignable to type '"timeseries" | undefined'.
+        type: 'linear',
+        time: {
+          // @ts-expect-error Type 'string' is not assignable to type 'false | "millisecond" | "second" | "minute" | "hour" | "day" | "week" | "month" | "quarter" | "year" | undefined'.
+          unit: 'year'
+        }
+      }
+    }
+  }
+});

--- a/types/tests/tsconfig.json
+++ b/types/tests/tsconfig.json
@@ -3,6 +3,7 @@
     "target": "ES6",
     "moduleResolution": "Node",
     "alwaysStrict": true,
+    "strict": true,
     "noEmit": true
   },
   "include": [

--- a/types/utils.d.ts
+++ b/types/utils.d.ts
@@ -12,7 +12,7 @@ export type DeepPartial<T> = T extends Function
   type _DeepPartialArray<T> = Array<DeepPartial<T>>
 type _DeepPartialObject<T> = { [P in keyof T]?: DeepPartial<T[P]> };
 
-export type DistributiveArray<T> = T extends unknown ? T[] : never
+export type DistributiveArray<T> = [T] extends [unknown] ? Array<T> : never
 
 // From https://stackoverflow.com/a/50375286
 export type UnionToIntersection<U> = (U extends any ? (k: U) => void : never) extends ((k: infer I) => void) ? I : never;

